### ISSUE TITLE
Include/allow upcoming epochs in epoch_stake

### ISSUE THREAD

**Issue by 1000101** [input-output-hk/cardano-db-sync#688]
Sorry if a similar issue has been filed, I haven't been able to find it so here goes.

Currently, there is no way to query active stake distribution for upcoming epoch(s).

The data should be already available, as active stake is taken from snapshot of live stakes from 2 epochs before. E.g. in the epoch 277, we already know active stake distribution for epoch 278 as the snapshot from epoch 276 has already been taken.

However, there is no way to query this data at the moment.

This could potentially even stretch to 2 upcoming epochs (once the computation of snapshot is done, although this would currently pose problems as `block_id` is present in `epoch_stake` table). But having at least one would be nice enough. Would this make sense / would you consider implementing this? Thanks!

**Comment by erikd** [comment]
This seems to be a duplicate of https://github.com/input-output-hk/cardano-db-sync/issues/599 .

**Comment by 1000101** [comment]
Thanks for the response. I've read through #599 but it's a different case. This issue is not about querying `live stake`, which is tricky and requires a lot of computation, but rather querying `active stake` of a not yet reached epoch.

Let's say we're currently in epoch 278. We already _know_ all the active stakes of upcoming epoch 279, but it's nowhere to be found in the db. It materializes only once we reach epoch 279.

**Comment by erikd** [comment]
> Let's say we're currently in epoch 278. We already know all the active stakes of upcoming epoch 279, but it's nowhere to be found in the db. It materializes only once we reach epoch 279.

Yes, I think the stake distribution for epoch `N + 1` should be available `k` slots (the stability period) after the start of epoch `N`. 

However, lower level code does not provide a nice API for notifying a client like `db-sync` of the availability of this data and IMO `db-sync` already reimplements far to much of the ledger logic for me to want to do more. Instead we are working towards the ledger notifying clients (like `db-sync`) of events (like this) which will make the whole system more robust. Currently the stake distribution for epoch `N` is only added to the database at the start of the epoch `N`. This is robust. Doing anything else would be duplicating logic in `db-sync` that is already in `ledger-specs`.

**Comment by Cmdv** [comment]
Based on the provided information, it seems that the issue has been addressed. We're closing this issue now. If you have any more concerns or if the problem reappears, please let us know by reopening this issue. We appreciate your help!
